### PR TITLE
implementation of fail-fast option for copy-from statement

### LINKED ISSUE #11205: [Feature Request] Fail-Fast option for COPY-FROM
**Use case**:
As a user I don't **always** want to wait till the whole `copy from` statement is complete, to see if there were any errors.

**Feature description**:
There is an `fail-fast` option for the `COPY FROM`, that if "at least one error" occured, the statement completes and i get the error in the `summary`  returned by `RETURNED SUMMARY`

### PR DESCRIPTION
**Pull request by jeeminso**
closes #11205

implement `fail-fast` option for `copy from` stmt. The current implementation uses `bulk_size` as a single unit of iteration. It makes sense to use this option in combination with `fail-fast` that every `bulk_size`, there will be a check for errors. 
 
 ex)
```
cr> copy tbl2 from '/home/jee/crate/t*.json' with(shared = false, fail_fast = true, bulk_size = 50000);                                                       
JobKilledException[Job killed. NODE: Pointe de Charbonnel
[URI: file:///home/jee/crate/t3.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=9646, line_numbers=[9, 13, 14, 24, 28, 39, 44, 68, 73, 88, 89, 93, 98, 103, 113, 134, 144, 149, 153, 168, 169, 173, 179, 193, 203, 234, 238, 249, 253, 259, 268, 278, 298, 299, 304, 309, 323, 338, 343, 354, 359, 369, 378, 399, 408, 424, 438, 444, 449, 453]}}],
[URI: file:///home/jee/crate/t6.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[8]}, failed to parse={count=1, line_numbers=[7]}}],
[URI: file:///home/jee/crate/t4.json, ERRORS: {}]
[URI: file:///home/jee/crate/t5.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[2]}, failed to parse={count=1, line_numbers=[1]}}]]

```

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: libs/sql-parser/src/main/antlr/SqlBase.g4 [hunk 1]
```diff
@@ -632,6 +633,11 @@ namedProperties
     : ident withProperties?
     ;
 
+copyFromOptions
+    : (RETURN SUMMARY)? (FAIL FAST)?
```
Comment (mfussenegger) [inline libs/sql-parser/src/main/antlr/SqlBase.g4:637]: Could we add an option for withProperties instead of going with new keywords?

Comment (jeeminso) [reply libs/sql-parser/src/main/antlr/SqlBase.g4]: Thanks Jordi, made the change.

File: docs/sql/statements/copy-from.rst [hunk 2]
```diff
@@ -372,6 +372,18 @@ specifies the size of one batch. The provided value must be greater than 0, the
...
+.. _sql-copy-from-fail_fast:
+
+``fail_fast``
+'''''''''''''
+
+With this option set to true, CrateDB will check for any errors at the end of
+each batch processing, specified by the
+:ref:`bulk_size <sql-copy-from-bulk_size>`. Any occurrences of the errors will
+trigger the execution to terminate that will prevent executing further while
+completing any in-flight executions normally.
```
Comment (mfussenegger) [inline docs/sql/statements/copy-from.rst:384]: I think this mentions too many implementation details.  This could be trimmed down to something like "If enabled, the COPY FROM operation will abort early after an error. This is best effort and due to the distributed execution, it may still continue processing some records before it aborts.

File: server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java [hunk 3]
```diff
@@ -155,7 +159,7 @@ private void consumeIterator() {
             return;
         }
         try {
-            while (batchIterator.moveNext()) {
+            while (!consumptionFinished && batchIterator.moveNext()) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:162]: Doesn't this have to take into account whether `failFast` is set? Otherwise this would generally stop processing records once an error occurs.

File: server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java [hunk 4]
```diff
@@ -110,7 +110,11 @@ public BatchIteratorBackpressureExecutor(UUID jobId,
 
     private void continueConsumptionOrFinish(@Nullable R result, Throwable failure) {
         if (result != null) {
-            resultRef.accumulateAndGet(result, combiner);
+            try {
+                resultRef.accumulateAndGet(result, combiner);
+            } catch (RuntimeException e) {
+                consumptionFinished = true;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:116]: Wouldn't it be enough to look for `failure` and abort the loop/processing if it is present and `failFast` is set. Without further extending the UpsertResultCollectors ?

**Review by mfussenegger — COMMENTED** [review]
I'm wondering if we should try to invoke a `KillRequest` on a failure, to ensure the operation is also interrupted on other nodes involved?

File: server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java (hunk 4 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:116]: my understanding is that `failure` is only meant for handling exceptions thrown from `batchIterator.loadNextBatch`. I instead need access to `errorCount` from `UpsertResults`. That was the reason why I had to extend `UpsertResultCollectors`.

File: server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java (hunk 3 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:162]: if `failFast` is set, the `UpsertResultCollector` implementation is wrapped with `FailFastResultCollector` which would throw and eventually set `consumptionFinished`.

File: docs/sql/statements/copy-from.rst (hunk 2 above)
Comment (jeeminso) [reply docs/sql/statements/copy-from.rst]: Thanks, this is very concise nothing more to add. :+1:

**Comment by jeeminso** [comment]
Hello @mfussenegger , I have made the doc change and replied to your comments. 
Overall, the choices I made were based on that the errors which trigger the abortion is only known/visible at result collection stage.

**Comment by jeeminso** [comment]
> I'm wondering if we should try to invoke a `KillRequest` on a failure, to ensure the operation is also interrupted on other nodes involved?

I had another look on my implementation, I do not think it is needed. My approach was to stop the batch iterator if any errors are observed during result collection.

below is my experiment to check if `killRequest` should be invoked:
```
// data to be inserted
$ head /t.json 
{"a":1234599}
{"a":123459999}
{"b":123}
{"b":1}
{"a":1234}
{"a":1234599}
{"a":123459999}
{"b":123}
{"b":1}
{"a":1234}
$ grep "a" /t.json | wc -l
14469
$ grep "b" /t.json | wc -l
9646

// 14469 successes and 9946 failures with fail_fast = false on both nodes
cr> copy tbl from '/t.json' with(shared = false, fail_fast = false, bulk_size = 1) return summary;                                                            
+-------------------------------------------------------------------+----------------+---------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| node                                                              | uri            | success_count | error_count | errors                                                                                                                                                                                                                                                                                                                                                      |
+-------------------------------------------------------------------+----------------+---------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {"id": "Ag_5JnJvS1WvJMkBI5TWUQ", "name": "Gross Düssi"}           | file:///t.json |         14469 |        9646 | {"mapping set to strict, dynamic introduction of [b] within [default] is not allowed": {"count": 9646, "line_numbers": [3, 9, 8, 13, 43, 63, 19, 54, 58, 69, 48, 44, 33, 74, 79, 149, 213, 178, 198, 189, 219, 234, 244, 248, 238, 49, 34, 104, 103, 118, 139, 174, 209, 224, 233, 88, 94, 99, 158, 128, 124, 144, 133, 164, 148, 199, 214, 188, 204, 68]}} |
| {"id": "r3X0eVfBTvm_COJQvUnaOQ", "name": "Hintere Stangenspitze"} | file:///t.json |         14469 |        9646 | {"mapping set to strict, dynamic introduction of [b] within [default] is not allowed": {"count": 9646, "line_numbers": [83, 93, 104, 138, 13, 29, 39, 49, 79, 33, 44, 69, 28, 3, 114, 173, 244, 24, 239, 178, 224, 199, 124, 119, 139, 148, 48, 64, 58, 228, 194, 103, 133, 164, 144, 84, 99, 89, 109, 129, 94, 208, 218, 4, 158, 14, 169, 183, 189, 229]}} |
+-------------------------------------------------------------------+----------------+---------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
COPY 2 rows in set (3.900 sec)  

// under 300 successes and 200 errors with fail_fast = true, on both nodes
cr> copy tbl from '/t.json' with(shared = false, fail_fast = true, bulk_size = 1) return summary;                                                             
+-------------------------------------------------------------------+----------------+---------------+-------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| node                                                              | uri            | success_count | error_count | errors                                                                                                                                                                                                                                                                                                                                             |
+-------------------------------------------------------------------+----------------+---------------+-------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {"id": "Ag_5JnJvS1WvJMkBI5TWUQ", "name": "Gross Düssi"}           | file:///t.json |           266 |         178 | {"mapping set to strict, dynamic introduction of [b] within [default] is not allowed": {"count": 178, "line_numbers": [54, 18, 39, 8, 14, 23, 43, 108, 4, 33, 84, 19, 73, 3, 79, 29, 178, 64, 83, 119, 128, 194, 209, 159, 88, 188, 203, 228, 249, 273, 9, 38, 59, 124, 148, 189, 103, 198, 169, 89, 158, 229, 94, 53, 179, 28, 63, 34, 69, 134]}} |
| {"id": "r3X0eVfBTvm_COJQvUnaOQ", "name": "Hintere Stangenspitze"} | file:///t.json |           113 |          76 | {"mapping set to strict, dynamic introduction of [b] within [default] is not allowed": {"count": 76, "line_numbers": [3, 8, 14, 28, 13, 23, 29, 68, 4, 133, 138, 38, 33, 178, 43, 59, 93, 103, 98, 63, 129, 49, 73, 88, 118, 124, 173, 128, 134, 148, 18, 48, 169, 44, 113, 58, 69, 114, 99, 159, 19, 64, 53, 9, 109, 104, 24, 183, 79, 108]}}     |
+-------------------------------------------------------------------+----------------+---------------+-------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
COPY 2 rows in set (0.553 sec)
```

**Comment by jeeminso** [comment]
@mfussenegger, Hello Jordi, the implementation passes simple manual tests and done initial refactoring. Please review the last commit to confirm overall direction. Thank you

**Review by mfussenegger — COMMENTED** [review]
Would it be possible to extend the existing kill mechanism with new properties instead and adjust the handling in the cases where the hard/soft difference is important?
We could introduce some `enum` or something to the existing `kill` methods and extend the `KillJobsRequest` as well.

The kill feature introduced quite a bit of complexity due to the fact that it usually involves two threads (one invoking the kill, another running an operation that should stop).
There are also some before/after conditions if the kill can happen before the operation even started and so on. Would be nice if we could avoid duplicating that kind of logic. (E.g. as done in the `CollectTask` here, or the second volatile read in the `FileReadingIterator`)

Testing it is also tricky due to the inherent non-deterministic nature of the problem, so it would be great if we didn't multiply the amount of scenarios we've to cover.

**Comment by jeeminso** [comment]
Thanks for reviewing @mfussenegger, I have made the changes and  since it is not suggested to add more fields to `FileReadingIterator` I instead embed it in the exception. Another change was to add `Type` parameter to `Killable.kill` but maybe I can remove it since it could be retreived from the exception now.

File: libs/shared/src/main/java/io/crate/data/Killable.java [hunk 5]
```diff
@@ -25,6 +25,11 @@
 
 public interface Killable {
 
+    enum Type {
+        HARD,
+        SOFT
+    }
+
```
Comment (mfussenegger) [inline libs/shared/src/main/java/io/crate/data/Killable.java:32]: I this these could use some docs to explain the difference in behavior

File: server/src/main/java/io/crate/exceptions/JobKilledException.java [hunk 6]
```diff
@@ -21,30 +21,44 @@
...
-    private JobKilledException(String reason) {
+    protected JobKilledException(String reason, Killable.Type killType) {
         super("Job killed. " + reason);
+        this.killType = killType;
     }
 
-    private JobKilledException() {
+    protected JobKilledException(Killable.Type killType) {
         super(MESSAGE);
+        this.killType = killType;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/exceptions/JobKilledException.java:61]: Why the permission change if there are also static factory methods?

File: server/src/main/java/io/crate/execution/jobs/kill/TransportSoftKillJobsNodeAction.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    protected CompletableFuture<Integer> doKill(SoftKillJobsRequest request) {
+        return tasksService.softKillJobs(request.toKill(), request.userName(), request.reason());
+    }
+
+    @Override
+    public SoftKillJobsRequest read(StreamInput in) throws IOException {
+        return new SoftKillJobsRequest(in);
+    }
+}
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/jobs/kill/TransportSoftKillJobsNodeAction.java:57]: Isn't it possible to use the existing request + transport?

File: libs/dex/src/main/java/io/crate/data/CollectingBatchIterator.java [hunk 8]
```diff
@@ -164,9 +164,11 @@ public boolean allLoaded() {
...
 
     @Override
-    public void kill(@Nonnull Throwable throwable) {
-        onKill.accept(throwable);
-        killed = throwable;
+    public void kill(@Nonnull Throwable throwable, Type killType) {
+        onKill.accept(throwable, killType);
+        if (killType == Type.HARD) {
+            killed = throwable;
+        }
```
Comment (mfussenegger) [inline libs/dex/src/main/java/io/crate/data/CollectingBatchIterator.java:171]: I think this would deserve a comment on why `killed` is only set in the `Type.HARD` case

File: libs/shared/src/main/java/io/crate/data/Killable.java (hunk 5 above)
Comment (mfussenegger) [reply libs/shared/src/main/java/io/crate/data/Killable.java:32]: I'm also not entirely sure about the naming here. I feel like they don't capture the semantics in a way that explains their difference. What's a soft kill and a hard kill?

If I understand the intention right, it is mostly about preserving parts of the results instead of failing with a "job killed". Maybe we could capture this somehow?

File: server/src/main/java/io/crate/exceptions/JobKilledException.java (hunk 6 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/exceptions/JobKilledException.java:61]: sorry for confusion, I tried extending it initially and not fully reverted afterwards. fixing now

File: libs/shared/src/main/java/io/crate/data/Killable.java (hunk 5 above)
Comment (jeeminso) [reply libs/shared/src/main/java/io/crate/data/Killable.java:32]: Thanks I agree, the intention was to make the target iterator or task to stall or do nothing, instead of throwing from it. I renamed it to `SUPPRESS_EXCEPTION`

File: libs/dex/src/main/java/io/crate/data/CollectingBatchIterator.java (hunk 8 above)
Comment (jeeminso) [reply libs/dex/src/main/java/io/crate/data/CollectingBatchIterator.java:171]: Thanks, added some explanation

**Comment by jeeminso** [comment]
Hello @mfussenegger, please have another look, Thank you

**Comment by jeeminso** [comment]
Thanks for reviewing and your suggestion, @mfussenegger. Before moving on to a complete different approach,
> Maybe we've to get back to the drawing board and consider a completely different approach for copy from in general? Maybe the existing BatchIterator pipeline is not the best fit?

I tried to generify as much as possible so that I can completely drop this approach if it turned out badly.

**Comment by jeeminso** [comment]
@mfussenegger Please review if the tight coupling has resolved. If the solution is feasible there are few things that I still need to address, Thank you.

**Comment by jeeminso** [comment]
@mfussenegger Hello Jordi, with the newly defined scope, the implementation became much simpler, please have a quick look at the implementation. Thanks

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 9]
```diff
@@ -396,13 +398,7 @@ private BufferedReader createBufferedReader(InputStream inputStream) throws IOEx
...
         boolean sharedStorage = Objects.requireNonNullElse(shared, fileInput.sharedStorageDefault());
         if (sharedStorage) {
-            moduloPredicate = input -> {
-                int hash = input.hashCode();
-                if (hash == Integer.MIN_VALUE) {
-                    hash = 0; // Math.abs(Integer.MIN_VALUE) == Integer.MIN_VALUE
-                }
-                return Math.abs(hash) % numReaders == readerNumber;
-            };
+            moduloPredicate = input -> moduloPredicateImpl.test(input, this.readerNumber, this.numReaders);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:401]: Why the changes here?

File: server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java [hunk 10]
```diff
@@ -101,6 +105,8 @@ public BatchIteratorBackpressureExecutor(UUID jobId,
         this.getDelayInMs = getDelayInMs;
         this.resultRef = new AtomicReference<>(identity);
         this.continueConsumptionOrFinish = this::continueConsumptionOrFinish;
+        this.earlyTerminationCondition = earlyTerminationCondition == null ? (results) -> false : earlyTerminationCondition;
+        this.resultsToFailure = resultsToFailure == null ? (result) -> new Throwable() : resultsToFailure;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:109]: Can we use a more specific exception than `Throwable`?

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java (hunk 9 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:401]: Ah I see it's used in the tests. Any reason to wrap it in a `TriPredicate` instead of creating a function with the `URI -> int -> int` signature directly?

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java]: no special reason, fixed it. Thank you

**Comment by jeeminso** [comment]
@mfussenegger Thanks for reviewing, I have addressed them and added minor message formatting.
Do you have any suggestion on the formatting?

```
JobKilledException[Job killed. NODE: {id=IS-m0zMzTEKTOQMT_jk03Q, name=Nabois Grande}
[URI: file:///home/jee/crate/t4.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[2]}, failed to parse={count=1, line_numbers=[1]}}]]
```

File: docs/sql/statements/copy-from.rst [hunk 11]
```diff
@@ -372,6 +372,16 @@ specifies the size of one batch. The provided value must be greater than 0, the
...
 
 
+.. _sql-copy-from-fail_fast:
+
+``fail_fast``
+'''''''''''''
+
+If enabled, the ``COPY FROM`` operation will abort early after an error. This
+is best effort and due to the distributed execution, it may still continue
+processing records before it aborts.
```
Comment (mfussenegger) [inline docs/sql/statements/copy-from.rst:382]:
```suggestion
A boolean value indicating if the ``COPY FROM`` operation should abort early after an error. This
is best effort and due to the distributed execution, it may continue
processing some records before it aborts.

Defaults to ``false``.
```

**Review by mfussenegger — APPROVED** [review]

**Comment by mfussenegger** [comment]
> @mfussenegger Thanks for reviewing, I have addressed them and added minor message formatting. Do you have any suggestion on the formatting?
> 
> ```
> JobKilledException[Job killed. NODE: {id=IS-m0zMzTEKTOQMT_jk03Q, name=Nabois Grande}
> [URI: file:///home/jee/crate/t4.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[2]}, failed to parse={count=1, line_numbers=[1]}}]]
> ```

I'd trim the node information to just the names or even omit them completely. Otherwise seems okay. The amount of errors reported shoudl stay close to 1, right?

**Comment by jeeminso** [comment]
> > @mfussenegger Thanks for reviewing, I have addressed them and added minor message formatting. Do you have any suggestion on the formatting?
> > ```
> > JobKilledException[Job killed. NODE: {id=IS-m0zMzTEKTOQMT_jk03Q, name=Nabois Grande}
> > [URI: file:///home/jee/crate/t4.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[2]}, failed to parse={count=1, line_numbers=[1]}}]]
> > ```
> 
> I'd trim the node information to just the names or even omit them completely. Otherwise seems okay. The amount of errors reported shoudl stay close to 1, right?

It will be all errors seen from a single bulk_size

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java [hunk 12]
```diff
@@ -226,15 +226,11 @@ public static ExecutionPlan planCopyFromExecution(AnalyzedCopyFrom copyFrom,
             clusteredByInputCol = InputColumns.create(table.getReference(clusteredBy), sourceSymbols);
         }
 
-        boolean failFast = boundedCopyFrom.settings().getAsBoolean("fail_fast", false);
-        if (failFast) { // fail_fast should be able to provide meaningful exception messages
-            copyFrom = new AnalyzedCopyFromReturnSummary(copyFrom.tableInfo(), copyFrom.table(), copyFrom.properties(), copyFrom.uri());
-        }
-
```
Comment (jeeminso) [inline server/src/main/java/io/crate/planner/statement/CopyFromPlan.java]: this was not enough, it only worked for failing cases. If no uri had errors, the operation should behave as if no summary was collected.

**Comment by jeeminso** [comment]
Hi @mfussenegger, I missed a case and now have a fix for it, would like another review. Thanks!

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java [hunk 13]
```diff
@@ -861,8 +862,8 @@ public void test_copy_from_with_fail_fast_property_with_one_node_configuration()
                          "data1.json",
                          target);
 
-
         execute("CREATE TABLE t (a int)");
+        ensureGreen();
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:866]: Why did the `ensureGreen` calls become necessary? They shouldn't be required

Comment (jeeminso) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:866]: I had a weird behaviour that resolved by having ensureGreen(), but seems gone. Taking all out and merging, Thank you.
